driver/compenv.ml

```diff
@@ -587,14 +631,75 @@ type deferred_action =
 let c_object_of_filename name =
   Filename.chop_suffix (Filename.basename name) ".c" ^ Config.ext_obj
 
+let check_ir name =
+  let check_suffix () =
+    let ext = Filename.extension name in
+    let ext_len = String.length ext in
+    if ext_len <= 0 then
+      None
+    else begin
+      List.find_opt (fun ir ->
+          let s = Compiler_ir.extension ir in
+          (* Check whether [ext] starts with [s]. It is not an exact match,
+             to distinguish different passes on the same IR. *)
+          let s_len = String.length s in
+          s_len <= ext_len && s = String.sub ext 0 s_len)
+        Compiler_ir.all
+    end
+  in
+  let check_magic () =
+    let ic = open_in_bin name in
+    Misc.try_finally
+      (fun () ->
+         (* Assumes that the length of magic numbers for all IRs is the same. *)
+         let len = String.length Config.linear_magic_number in
+         let pre_len = len - 3 in
+         try
+           let buffer = really_input_string ic len in
+           List.find_opt (fun ir ->
+             let magic = Compiler_ir.magic ir in
+             assert (String.length magic = len);
+             if String.equal buffer magic then true
+             else begin
+               if String.sub buffer 0 pre_len = String.sub magic 0 pre_len
+               then Misc.fatal_errorf
+                      "%s is not compiled for this version of OCaml." name ()
+               else false
+             end)
+             Compiler_ir.all
+         with End_of_file -> None
```

Comment:
That seems a bit unusual to me. I believe that usually, we only analyse the extension to decide what the file is and then check the magic number later on. Maybe analysing the file contents is fine, but I feel like this should be a separate discussion. For instance, it means that we are opening an input file twice, which I don't think we do otherwise. Perhaps such an analysis should also be centralised so that we recognise all kinds of files produced by the compiler (cmi, ast files, ...).

Reply:
(I haven't followed the present PR, apologies, just reacting to Jérémie's comment here.)

I have a PR in-review that implements centralized logic for magic-number reading (not other aspect of file-format analysis): #463 . I don't mean to couple the two PRs, but maybe this suggests that this particular part of the present PR could be improved later on, when #463 will be merged.

Reply:
@diml  You are right, it is not necessary to check for magic numbers here. 

I added it to match the behavior in the frontend, where an input file that contains a parsetree can be recognised by its magic number, not the extension. I do not see a nice way to do it for IR files without opening the file twice. Note that it will not affect normal compilation, as the check for IR will only be done when in the previous version the compiler would have terminated with an error about an unknown file.   It sounds from your comment that the cost of opening the file twice is too high, or perhaps there are other bad consequences. I can remove it and maybe we reconsider it after the infrastructure of #463 is in place.

Reply:
I see. Note that for the parsetree it's a bit different as we expect the file to contain a certain type of value and only check that it contains what we expect. The magic number stuff is just to decide whether the data is in text form or in marshaled form. But whether the file is supposed to be a structure or signature is decided by the extension or by passing `-impl <file>` or `-intf <file>`. Passing a marshaled structure to `-intf` or in a file with the `.mli` extension would cause an error for instance.

Here it is a bit different as we decide what to do based on the contents of the file. Maybe that's ok, but it feels like it should be a separate discussion.

Reply:
Ok, thanks for the explanation! I've removed the logic that checks magic numbers here.